### PR TITLE
[BEAM-9650] Cleanup documentation on side inputs patterns

### PR DESCRIPTION
**Pull request by Ardagan**
[[Website staging preview]](http://apache-beam-website-pull-requests.storage.googleapis.com/11415/documentation/patterns/side-inputs/index.html)
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by Ardagan** [comment]
R: @soyrice

File: website/src/documentation/patterns/side-inputs.md [hunk 1]
```diff
@@ -50,25 +50,24 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
 
 ## Slowly updating side input using windowing
 
-You can read side input pcollection periodically into distinct windows.
-Later, when you apply side input to your main input, windows will be matched automatically 1:1.
-This way, you can guarantee side input consistency on the duration of the single window.
+You can read side input data periodically into distinct PCollection windows.
```
Comment (Ardagan) [inline website/src/documentation/patterns/side-inputs.md:53]: In Beam, we have IO.Read transforms that do read source data. Later data read can be assigned to specific window.

File: website/src/documentation/patterns/side-inputs.md [hunk 2]
```diff
@@ -50,25 +50,25 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
-To do this, you can utilize PeriodicSequence PTransform that will generate infinite sequence
-of elements with some real-time period:
+You can read side input data periodically into distinct PCollection windows.
+Later, when you apply the side input to your main input, each main input
+window is automatically matched to a single side input window.
+This guarantees side input consistency on the duration of the single window,
+meaning that each window on main input side will be matched to a single
+version of side input data.
+
+To do this, you can utilize a combination of PeriodicSequence/PeriodicImpulse
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:60]: For formatting, it might be clearer to say "a combination of PeriodicSequence and PeriodicImpulse PTransforms" instead of "PeriodicSequence/PeriodicImpulse PTransforms" - just to make sure there's no ambiguity in regards to these being two separate transforms.

File: website/src/documentation/patterns/side-inputs.md [hunk 3]
```diff
@@ -50,25 +50,25 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+PTransforms that will generate infinite sequence of elements with some real-time
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:61]: Present tense -> "PTransforms that generate"

Missing article -> "generate an infinite sequence..."

File: website/src/documentation/patterns/side-inputs.md [hunk 4]
```diff
@@ -50,25 +50,25 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+period and SDF Read operation or similar to read data into side input
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:62]: I think we should split this sentence into two - it's hard to follow. A good place to start a new sentence is at "SDF Read operation"

File: website/src/documentation/patterns/side-inputs.md [hunk 5]
```diff
@@ -50,25 +50,25 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+periodically:
 
 1. Use the PeriodicImpulse transform to generate windowed periodic sequence.
 
-    a. MAX_TIMESTAMP can be replaced with some closer boundary if you want to stop generating elements at some point.
-
 1. Read data using Read operation triggered by arrival of PCollection element.
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:67]: On a second look, I think it's redundant to say "Read data using a Read operation" - maybe we can say something like "Read data when a PCollection element arrives"

File: website/src/documentation/patterns/side-inputs.md [hunk 6]
```diff
@@ -50,25 +50,25 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
 
 1. Apply side input.
 
-```python
+```py
```
Comment (Ardagan) [inline website/src/documentation/patterns/side-inputs.md:71]: Seems that adding this snippet makes java snippet at L45 disappear from served site despite me not changing anything in code above. Do you know if there's some trick to make both snippets appear?

Comment (Ardagan) [reply website/src/documentation/patterns/side-inputs.md:71]: Website sets config for the page to show only python or java snippets. It seem to pick the last language provided as default.

File: website/src/documentation/patterns/side-inputs.md [hunk 7]
```diff
@@ -46,29 +54,36 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
-To do this, you can utilize PeriodicSequence PTransform that will generate infinite sequence
-of elements with some real-time period:
+Described approach can be implemented using combination of
+PeriodicSequence or PeriodicImpulse PTransforms and SDF Read or ReadAll
+PTransforms:
 
-1. Use the PeriodicImpulse transform to generate windowed periodic sequence.
-
-    a. MAX_TIMESTAMP can be replaced with some closer boundary if you want to stop generating elements at some point.
+1. Use the PeriodicImpulse or PeriodicSequence PTransform to generate an
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:75]: Oh! Can we format this with bullet points:

1. Use the PeriodicImpulse... to:

* Generate an infinite sequence of elements...
* Assign the elements to separate windows...

1. Read data...

File: website/src/documentation/patterns/side-inputs.md [hunk 8]
```diff
@@ -46,29 +54,36 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+You can read side input data periodically into distinct PCollection windows.
+Later, when you apply the side input to your main input, each main input
+window is automatically matched to a single side input window.
+This guarantees consistency on the duration of the single window,
+meaning that each window on the main input will be matched to a single
+version of side input data.
 
-To do this, you can utilize PeriodicSequence PTransform that will generate infinite sequence
-of elements with some real-time period:
+Described approach can be implemented using combination of
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:71]: How about something like this: "To read side input data periodically... use a combination of PeriodicSequence..."

File: website/src/documentation/patterns/side-inputs.md [hunk 9]
```diff
@@ -46,29 +54,36 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+PeriodicSequence or PeriodicImpulse PTransforms and SDF Read or ReadAll
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:72]: SDF Read and ReadAll are transforms too, right? The following will help clarify what the combination consists of: "...PeriodicSequence or PeriodicImpulse PTransforms and SDF Read or ReadAll PTransforms"

File: website/src/documentation/patterns/side-inputs.md [hunk 10]
```diff
@@ -46,29 +54,36 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+PTransforms:
 
-1. Use the PeriodicImpulse transform to generate windowed periodic sequence.
-
-    a. MAX_TIMESTAMP can be replaced with some closer boundary if you want to stop generating elements at some point.
+1. Use the PeriodicImpulse or PeriodicSequence PTransform to generate an
+infinite sequence of elements at required processing time intervals and assign
+them to separate windows.
 
 1. Read data using Read operation triggered by arrival of PCollection element.
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:79]: I see you marked the previous comment here as resolved but I don't see any changes to this line. Was that intended?

File: website/src/documentation/patterns/side-inputs.md (hunk 9 above)
Comment (Ardagan) [reply website/src/documentation/patterns/side-inputs.md:72]: I might be missing something, but that's exactly the phrasing. "PTransforms" word is on next line.

File: website/src/documentation/patterns/side-inputs.md (hunk 8 above)
Comment (Ardagan) [reply website/src/documentation/patterns/side-inputs.md]: I'm a bit confused here since adding text you suggest is pretty much re-writing same thing that is a header and previous abstract. In next iteration I tried to remove this abstract overall since list of steps provides same information.

File: website/src/documentation/patterns/side-inputs.md (hunk 10 above)
Comment (Ardagan) [reply website/src/documentation/patterns/side-inputs.md]: Was not intended. FIxed.

File: website/src/documentation/patterns/side-inputs.md (hunk 8 above)
Comment (soyrice) [reply website/src/documentation/patterns/side-inputs.md]: Gotcha - the intent is to summarize the previous abstract, so that it's easier for readers to skim the doc.

File: website/src/documentation/patterns/side-inputs.md (hunk 6 above)
Comment (soyrice) [reply website/src/documentation/patterns/side-inputs.md:84]: Yeah, the issue is that the page can only show one language at a time.

File: website/src/documentation/patterns/side-inputs.md [hunk 11]
```diff
@@ -46,29 +54,34 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
-1. Read data using Read operation triggered by arrival of PCollection element.
-
+You can read side input data periodically into distinct PCollection windows.
+Later, when you apply the side input to your main input, each main input
+window is automatically matched to a single side input window.
+This guarantees consistency on the duration of the single window,
+meaning that each window on the main input will be matched to a single
+version of side input data.
+
+Implementation of described approach can be narrowed down to:
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:71]: We should rephrase as "To read side input data periodically:"

The goal is to summarize the previous paragraph, so that it's clear what these steps refer to without needing to read through the previous paragraph. The previous paragraph offers additional context, but if a reader skims the page, they should be able to quickly figure out what these steps are about.

File: website/src/documentation/patterns/side-inputs.md [hunk 12]
```diff
@@ -46,29 +54,34 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
+
+1. Use the PeriodicImpulse or PeriodicSequence PTransform to: 
+    * Generate an infinite sequence of elements at required processing time
+    intervals
+    * Assign them to separate windows.
+1. Fetch data using SDF Read or ReadAll PTransform triggered by arrival of
+PCollection element.
 1. Apply side input.
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:79]: Missing "the" -> "Apply the side input"

**Comment by soyrice** [comment]
> The text LGTM after resolving the remaining comments, but I don't see the code sample on the staging server.

I think it's missing because it's not picking up the language selector.

**Comment by Ardagan** [comment]
Staging job failing together with website publish jobs :( So staging version has old code unfortunately.

File: website/src/documentation/patterns/side-inputs.md (hunk 9 above)
Comment (soyrice) [reply website/src/documentation/patterns/side-inputs.md]: I probably just missed the word. Looks good in staging.

File: website/src/documentation/patterns/side-inputs.md [hunk 13]
```diff
@@ -46,29 +54,34 @@ For instance, the following code sample uses a `Map` to create a `DoFn`. The `Ma
...
-This way, you can guarantee side input consistency on the duration of the single window.
-
-To do this, you can utilize PeriodicSequence PTransform that will generate infinite sequence
-of elements with some real-time period:
-
-1. Use the PeriodicImpulse transform to generate windowed periodic sequence.
-
-    a. MAX_TIMESTAMP can be replaced with some closer boundary if you want to stop generating elements at some point.
+You can read side input data periodically into distinct PCollection windows.
+Later, when you apply the side input to your main input, each main input
```
Comment (soyrice) [inline website/src/documentation/patterns/side-inputs.md:65]: I think we should remove "Later" because this is a part of the overall workflow described in the previous sentence, rather than a secondary step.

So this should be: "When you apply the side input to your main input..."

**Comment by soyrice** [comment]
LGTM (modulo one last comment). Thanks again for opening this PR and doing an editorial review
